### PR TITLE
Clean up Windows RNG

### PR DESCRIPTION
**Pull request by xoloki**
This PR removes all of the dangerous Windows entropy gathering routines in favor of standard CryptGenRandom calls, as was discussed in the "Improving OpenSSL default RNG" thread on openssl-dev. This fixes common, repeatable crashes that happen when running openssl under the VS debugger.

### CODE REVIEW

**Comment by xoloki** [comment]
nmake test passes

**Comment by xoloki** [comment]
@mattcaswell is this PR okay?

File: crypto/rand/rand_win.c [hunk 1]
```diff
@@ -567,51 +166,6 @@ int RAND_poll(void)
     return (1);
 }
 
-int RAND_event(UINT iMsg, WPARAM wParam, LPARAM lParam)
```
Comment (mattcaswell) [inline crypto/rand/rand_win.c]: I'm wondering whether we should retain RAND_event(), and just have it call RAND_poll() (ignoring all params), and return RAND_status()? i.e. for backwards compat.

File: crypto/rand/rand_win.c [hunk 2]
```diff
@@ -567,51 +166,6 @@ int RAND_poll(void)
...
-
-    readtimer();
-    RAND_add(&iMsg, sizeof(iMsg), add_entropy);
-    RAND_add(&wParam, sizeof(wParam), 0);
-    RAND_add(&lParam, sizeof(lParam), 0);
-
-    return (RAND_status());
-}
-
-void RAND_screen(void)
```
Comment (mattcaswell) [inline crypto/rand/rand_win.c]: Similarly for RAND_screen(). Perhaps we should retain, but remove the "readscreen()" call below.

File: include/openssl/rand.h [hunk 3]
```diff
@@ -113,13 +113,6 @@ int RAND_egd_bytes(const char *path, int bytes);
 # endif
 int RAND_poll(void);
 
-# if defined(OPENSSL_SYS_WINDOWS) || defined(OPENSSL_SYS_WIN32)
-
-void RAND_screen(void);
-int RAND_event(UINT, WPARAM, LPARAM);
```
Comment (mattcaswell) [inline include/openssl/rand.h]: ...and if we retain them then these would be wrapped in DEPRECATEDIN macros.

**Comment by mattcaswell** [comment]
Yes, this PR looks a lot better. I just have a few minor comments.

File: crypto/rand/rand_win.c (hunk 1 above)
Comment (richsalz) [inline crypto/rand/rand_win.c]: Via ```OPENSSL_API_COMPAT``` ?

**Comment by mattcaswell** [comment]
For the sake of maintaining the history, this PR replaces #512.

File: crypto/rand/rand_win.c (hunk 1 above)
Comment (xoloki) [inline crypto/rand/rand_win.c]: I think this is a good time to make a clean break.  But keeping them as deprecated would be a reasonable option.

Comment (mattcaswell) [inline crypto/rand/rand_win.c]: Yes, I think we keep them but deprecated (using the OPENSSL_API_COMPAT macro as @richsalz suggests), and all they do is call RAND_poll/RAND_status.

Comment (xoloki) [inline crypto/rand/rand_win.c]: So 

if OPENSSL_API_COMPAT < 0x10100000L

?

Comment (mattcaswell) [inline crypto/rand/rand_win.c]: Yes. And wrap the header file declarations in DEPRECATEDIN_1_1_0()

Comment (mattcaswell) [inline crypto/rand/rand_win.c]: Oh...which also means the documentation needs to change too :-)

**Comment by xoloki** [comment]
Ok, functions added back as deprecated in header/source/docs.  nmake test still passes.

File: doc/crypto/RAND_add.pod [hunk 4]
```diff
@@ -42,25 +42,15 @@ or L<RAND_load_file(3)>.
...
-randomness. It should be noted that both methods cannot be used on
-servers that run without user interaction.
-
 =head1 RETURN VALUES
 
-RAND_status() and RAND_event() return 1 if the PRNG has been seeded
+RAND_status() returns 1 if the PRNG has been seeded
 with enough data, 0 otherwise.
 
+RAND_event() calls RAND_poll() and returns RAND_seed().
```
Comment (mattcaswell) [inline doc/crypto/RAND_add.pod]: No, it returns RAND_status().

File: doc/crypto/RAND_add.pod [hunk 5]
```diff
@@ -42,25 +42,15 @@ or L<RAND_load_file(3)>.
...
+
+RAND_screen calls RAND_poll().
+
 The other functions do not return values.
 
 =head1 SEE ALSO
```
Comment (mattcaswell) [inline doc/crypto/RAND_add.pod]: Please can you add an HISTORY section stating that RAND_screen() and RAND_event() were deprecated in 1.1.0 (see other pod files for an example).

File: doc/crypto/RAND_add.pod [hunk 6]
```diff
@@ -42,25 +42,15 @@ or L<RAND_load_file(3)>.
...
-event message (if any), and add it to the PRNG. The program can then
-process the messages as usual.
-
-The RAND_screen() function is available for the convenience of Windows
-programmers. It adds the current contents of the screen to the PRNG.
-For applications that can catch Windows events, seeding the PRNG by
-calling RAND_event() is a significantly better source of
-randomness. It should be noted that both methods cannot be used on
-servers that run without user interaction.
-
```
Comment (mattcaswell) [inline doc/crypto/RAND_add.pod]: Just leave a one liner in here for RAND_event() and RAND_screen() saying that they are deprecated and should not be called. Also in the synopsis section, move the declarations to the bottom and include the OPENSSL_API_COMPAT bit (like you did in rand.pod)

Comment (xoloki) [inline doc/crypto/RAND_add.pod]: Done, please check.

File: doc/crypto/RAND_add.pod [hunk 7]
```diff
@@ -42,27 +44,24 @@ or L<RAND_load_file(3)>.
...
+RAND_event() calls RAND_poll() and returns RAND_status().
+
+RAND_screen calls RAND_poll().
+
 The other functions do not return values.
 
+=head1 HISTORY
+
+RAND_event() and RAND_screen() are deprecated since OpenSSL
+1.1.  Use the functions described above instead.
```
Comment (mattcaswell) [inline doc/crypto/RAND_add.pod]: Nit: Should be 1.1.0

Comment (xoloki) [inline doc/crypto/RAND_add.pod]: Done.

**Comment by mattcaswell** [comment]
Just done a test build and everything looks good. One final thing I spotted. libcrypto.num needs updating to mark RAND_event() and RAND_screen() as DEPRECATEDIN_1_1_0

**Comment by xoloki** [comment]
Done @mattcaswell

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
Needs a second team review now.

**Comment by mattcaswell** [comment]
Ping anyone on the team?

**Comment by metaskills** [comment]
Definitely watching this as I think it relates to https://github.com/rails-sqlserver/tiny_tds/issues/250

**Comment by xoloki** [comment]
Agreed @metaskills, the bug referenced is clearly the result of the heap walk code which this PR removes.  Just need another team review before this PR can be merged.

**Comment by levitte** [comment]
I've been eyeing this off an on, and it looks sane to me, so +1

**Comment by xoloki** [comment]
Thanks @levitte!

**Comment by matbech** [comment]
You may want to consider using RtlGenRandom instead of CryptGenRandom. While the MSDN documentation says that RtlGenRandom https://msdn.microsoft.com/en-us/library/windows/desktop/aa387694(v=vs.85).aspx is deprecated, in reality this is not true. For example the rand implementation in the latest Microsoft CRT uses RtlGenRandom. The benefit is that it is simpler to use: 
- no crypto context must be acquired (CryptAcquireContext). No locks are necessary in case the crypto context is reused.
- RtlGenRandom never fails

CryptGenRandom is actually just a wrapper around RtlGenRandom but with the CryptAcquireContext overhead.

**Comment by xoloki** [comment]
I'd argue against using the deprecated interface, that's kind of the whole point of this PR.  And I really think we should acquire locks, that's what caused the crashes in the old heap walk code.

**Comment by thomthom** [comment]
> While the MSDN documentation says that RtlGenRandom https://msdn.microsoft.com/en-us/library/windows/desktop/aa387694(v=vs.85).aspx is deprecated, in reality this is not true. For example the rand implementation in the latest Microsoft CRT uses RtlGenRandom. 

Because the CRT uses it doesn't mean it's not deprecated. The fact they added that deprecation mark is a hint they want people moving away from it. Just because some other part of MS still use it doesn't mean it's the best future wise.

This CL replaces uses of old functions that's been deprecated - doesn't sound like a great idea to swap them out for other deprecated functions.

Better right, than easy, IMO.

From the MSDN article linked:

> The RtlGenRandom function is available for use in the operating systems specified in the Requirements section. It may be altered or unavailable in subsequent versions. Instead, use the CryptGenRandom function.

The important part of that is _"It may be altered or unavailable in subsequent versions"_  which should be a big warning to not use the function. Especially for a popular library such as OpenSSL that is so widely used. 

If you use an API that isn't assured to be around in the future it's just causing problems up stream.

**Comment by matbech** [comment]
For reference, the same issue has been raised in Go:
https://github.com/golang/go/issues/15589

Firefox has asked Microsoft for clarification a while ago:
https://bugzilla.mozilla.org/show_bug.cgi?id=504270

If the argument is not to use "deprecated" functions, at least consider using BCryptGenRandom on Windows 7 and higher as CryptoAPI itself is a legacy API (superseded by NCrypt/BCrypt).

Also for reference, the following projects have decided against the CryptoAPI:
- rand in all Microsoft CRT implementations (RtlGenRandom)
- .NET (BCryptGenRandom)
- Chromium (RtlGenRandom)
- Go (pending patch to use RtlGenRandom)
- Firefox (RtlGenRandom)

It's clear that CryptGenRandom has an unnecessary overhead which can be avoided by either using BCryptGenRandom or RtlGenRandom.

**Comment by xoloki** [comment]
Another part of this PR was removing all the #ifdef code for various windows platforms.  I don't think adding another for Win7 and greater is a good fit for this change.

CryptoAPI may be legacy, but it's not deprecated, so I think it's the appropriate choice here.  We're only grabbing 32 bytes from it anyway.  If this was being called all the time then tiny changes in performance would be worth increasing code complexity.  But I just don't see the tradeoff here.

**Comment by richsalz** [comment]
Agree with @xoloki.  It's part of the official API, this cleans up things a great deal, and is portable across all Windows platforms.  Perhaps in the next release, when we drop XP and other old platforms, we can look at BCryptoGenRandom.

Also, doesn't this address a couple of open RT tickets?

+1

**Comment by t-j-h** [comment]
+1 in its current form ...

**Comment by richsalz** [comment]
So, want me to merge?

**Comment by richsalz** [comment]
thanks for your efforts!  merged!!

**Comment by xoloki** [comment]
Thanks so much @mattcaswell, @levitte, and @richsalz!

**Comment by viniciusgati** [comment]
any chance to have this causing problems on linux?

**Comment by mattcaswell** [comment]
This is all in code that doesn't even get compiled on linux, so there should be no impacts.

File: crypto/rand/rand_win.c (hunk 1 above)
Comment (gvanem) [reply crypto/rand/rand_win.c]: > Oh...which also means the documentation needs to change too :-)
I noted that too here:
https://github.com/openssl/openssl/blob/1dc1ea182be183d8a393fdce4494360aee059cd2/NOTES-Windows.txt#L167

This change means `gdi32.dll` is no longer needed?

Comment (mspncp) [reply crypto/rand/rand_win.c]: _Note: See also this thread on [openssl-users](https://mta.openssl.org/pipermail/openssl-users/2020-July/012800.html)._

**Pull request closed without merging**